rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/rascal/RascalTextDocumentService.java

```diff
@@ -415,6 +428,50 @@ public CompletableFuture<List<? extends CodeLens>> codeLens(CodeLensParams param
             ;
     }
 
+    @Override
+    public CompletableFuture<List<Either<Command, CodeAction>>> codeAction(CodeActionParams params) {
+        logger.debug("codeAction: {}", params);
+
+        final var loc = Locations.toLoc(params.getTextDocument());
+        final var start = params.getRange().getStart();
+        // convert to Rascal 1-based line
+        final var startLine = start.getLine() + 1;
+        // convert to Rascal UTF-32 column width
+        final var startColumn = columns.get(loc).translateInverseColumn(start.getLine(), start.getCharacter(), false);
+
+        // first we make a future stream for filtering out the "fixes" that were optionally sent along with earlier diagnostics
+        // and which came back with the codeAction's list of relevant (in scope) diagnostics:
+        // CompletableFuture<Stream<IValue>>
+        CompletableFuture<Stream<IValue>> quickfixes
+            = CodeActions.extractActionsFromDiagnostics(params, rascalServices::parseCodeActions);
+
+        // here we dynamically ask the contributions for more actions,
+        // based on the cursor position in the file and the current parse tree
+        CompletableFuture<Stream<IValue>> codeActions = recoverExceptions(
+            getFile(params.getTextDocument())
+                .getCurrentTreeAsync()
+                .thenApply(Versioned::get)
+                .thenCompose((ITree tree) -> computeCodeActions(startLine, startColumn, tree, facts.getPathConfig(loc)))
+                .thenApply(IList::stream)
+            , () -> Stream.<IValue>empty())
+            ;
+
+        // final merging the two streams of commmands, and their conversion to LSP Command data-type
+        return CodeActions.mergeAndConvertCodeActions(this, "", BaseWorkspaceService.RASCAL_LANGUAGE, quickfixes, codeActions);
+    }
+
+    private CompletableFuture<IList> computeCodeActions(final int startLine, final int startColumn, ITree tree, PathConfig pcfg) {
+        IList focus = TreeSearch.computeFocusList(tree, startLine, startColumn);
```

Comment:
please move this calculation of the focus list into the completeable future, to move as much of the calculation to async task.

Reply:
this code was already called from the asynhronous task on lin 452. but I created yet another async wrapper anyway. Don't know if that helps.